algorithms/src/snark/marlin/ahp/prover/round_functions/fourth.rs

```diff
@@ -12,58 +12,229 @@
+            assert_eq!(circuit_a, circuit_c);
+            let (sum_a, lhs_a, g_a) = results_a?;
+            let (sum_b, lhs_b, g_b) = results_b?;
+            let (sum_c, lhs_c, g_c) = results_c?;
+            let matrix_sum = prover::message::MatrixSums { sum_a, sum_b, sum_c };
+            sums.push(matrix_sum);
+            state.circuit_specific_states.get_mut(circuit_a).unwrap().lhs_polynomials = Some([lhs_a, lhs_b, lhs_c]);
+            let matrix_gs = prover::MatrixGs { g_a, g_b, g_c };
+            gs.insert(circuit_a.id, matrix_gs);
+        }
+
+        let msg = prover::FourthMessage { sums };
+        let oracles = prover::FourthOracles { gs };
+
+        assert!(oracles.matches_info(&Self::fourth_round_polynomial_info(
+            state.circuit_specific_states.keys().map(|c| (c.id, &c.index_info))
+        )));
+
+        end_timer!(round_time);
+
+        Ok((msg, oracles, state))
     }
 
-    /// Output the degree bounds of oracles in the third round.
-    pub fn fourth_round_polynomial_info() -> BTreeMap<PolynomialLabel, PolynomialInfo> {
-        [("h_2".into(), PolynomialInfo::new("h_2".into(), None, None))].into()
+    #[allow(clippy::too_many_arguments)]
+    fn calc_matrix_sumcheck_witness(
+        label: String,
+        constraint_domain: EvaluationDomain<F>,
+        variable_domain: EvaluationDomain<F>,
+        non_zero_domain: EvaluationDomain<F>,
+        arithmetization: &MatrixArithmetization<F>,
+        alpha: F,
+        beta: F,
+        v_R_i_alpha_v_C_i_beta: F,
+        max_non_zero_domain: EvaluationDomain<F>,
+        fft_precomputation: &FFTPrecomputation<F>,
+        ifft_precomputation: &IFFTPrecomputation<F>,
+    ) -> Result<(Sum<F>, Lhs<F>, Gpoly<F>)> {
+        let mut job_pool = snarkvm_utilities::ExecutionPool::with_capacity(2);
+        job_pool.add_job(|| {
+            let a_poly_time = start_timer!(|| format!("Computing a poly for {label}"));
+            let a_poly = {
+                let coeffs = cfg_iter!(arithmetization.row_col_val.as_dense().unwrap().coeffs())
+                    .map(|v| v_R_i_alpha_v_C_i_beta * v)
+                    .collect();
+                DensePolynomial::from_coefficients_vec(coeffs)
+            };
+            end_timer!(a_poly_time);
+            a_poly
+        });
+
+        let (row_on_K, col_on_K, row_col_on_K) =
+            (&arithmetization.evals_on_K.row, &arithmetization.evals_on_K.col, &arithmetization.evals_on_K.row_col);
+        let R_size = constraint_domain.size_as_field_element;
+        let C_size = variable_domain.size_as_field_element;
+
+        job_pool.add_job(|| {
+            let b_poly_time = start_timer!(|| format!("Computing b poly for {label}"));
+            let alpha_beta = alpha * beta;
+            let b_poly = {
+                let evals: Vec<F> = cfg_iter!(row_on_K.evaluations)
+                    .zip_eq(&col_on_K.evaluations)
+                    .zip_eq(&row_col_on_K.evaluations)
```

Comment:
I think we should omit `row_col_on_K.evaluations` from the proving key, as it can be obtained by computing `r_c = r * c`. This should help reduce the proving key size nicely. (We should still keep `row_col` in the verifying key, of course.

Reply:
Nice 2bdf9248e

Depending on our existing metrics infra, In the future I will add tests to check the pk and vk size to avoid regressions.